### PR TITLE
Add tests for rake task to update versions file

### PR DESCRIPTION
**Pull request by sonalkr132**
closes: #1456 

Unfortunately this is going to fail because rake task is still incorrect :tongue: I would propose that instead of [removing the sql query](https://github.com/rubygems/rubygems.org/pull/1453/files#diff-2cf34b93c3257e00201d866038d19583L99), it should be corrected?

cc: @segiddins

### CODE REVIEW

**Comment by segiddins** [comment]
I'll look at this over the weekend and PR more test cases

**Comment by sonalkr132** [comment]
@segiddins Did you have time to look into this? :pray:

**Comment by segiddins** [comment]
😔 totally lost track. Will review this week. Sorry.

**Comment by segiddins** [comment]
```diff
diff --git a/lib/tasks/compact_index.rake b/lib/tasks/compact_index.rake
index c956dea..954e2a2 100644
--- a/lib/tasks/compact_index.rake
+++ b/lib/tasks/compact_index.rake
@@ -86,7 +86,10 @@ namespace :compact_index do
       info_checksum = versions.last.info_checksum
 
       # Only indexed versions of a gem go into *new* versions file
-      versions.reject! { |v| v.number.start_with?("-") }
+      yanked_versions = versions.select { |v| v.number.start_with?("-") }
+      versions -= yanked_versions
+      yanked_versions.map! { |v| v.number.sub(/^-/, '') }
+      versions.reject! { |v| yanked_versions.include?(v.number) }
 
       # Set all versions' info_checksum to work around https://github.com/bundler/compact_index/pull/20
       versions.each { |v| v.info_checksum = info_checksum }
diff --git a/test/unit/update_versions_file_test.rb b/test/unit/update_versions_file_test.rb
index a6bc843..4713b4e 100644
--- a/test/unit/update_versions_file_test.rb
+++ b/test/unit/update_versions_file_test.rb
@@ -71,15 +71,17 @@ class UpdateVersionsFileTest < ActiveSupport::TestCase
       setup do
         create(:version,
           rubygem:       @rubygem,
-          created_at:    2.minutes.ago,
+          created_at:    5.minutes.ago,
           number:        "0.0.1",
           info_checksum: "qw212r")
         create(:version,
-          indexed: false,
+          indexed:       false,
           rubygem:       @rubygem,
+          created_at:    3.minutes.ago,
           yanked_at:     1.minute.ago,
           number:        "0.0.2",
-          info_checksum: "sd12q")
+          info_checksum: "sd12q",
+          yanked_info_checksum: "qw212r")
         Rake::Task["compact_index:update_versions_file"].invoke
       end
 
```

**Comment by sonalkr132** [comment]
```diff
+      yanked_versions = versions.select { |v| v.number.start_with?("-") }
+      versions -= yanked_versions
+      yanked_versions.map! { |v| v.number.sub(/^-/, '') }
+      versions.reject! { |v| yanked_versions.include?(v.number) }
```

Instead, I have refactored [`versions`](https://github.com/bundler/bundler-api/blob/02c2633278edc81bf2e58bf9d7022b8607fd02e5/lib/bundler_api/gem_info.rb#L69) implementation using [remove control flag](http://martinfowler.com/bliki/FlagArgument.html) to avoid [duplication previously needed](https://github.com/rubygems/rubygems.org/pull/1454/files#diff-2cf34b93c3257e00201d866038d19583L76). Also, note that when we are creating new file, we want to use `version`'s [`checksum/sha256`](https://github.com/bundler/bundler-api/blob/02c2633278edc81bf2e58bf9d7022b8607fd02e5/lib/bundler_api/gem_info.rb#L86) for creating `info_checksum`. [`entry[:checksum]`](https://github.com/bundler/bundler-api/blob/02c2633278edc81bf2e58bf9d7022b8607fd02e5/lib/bundler_api/gem_info.rb#L99) was `nil` when we asked for both indexed and yanked versions. I haved added patch to match the behavior of bundler-api.
Let me know what you think.

**Comment by segiddins** [comment]
There's an issue I see with that last commit, let me write up a test

**Comment by sonalkr132** [comment]
Please enjoy this gif while @segiddins gets back the test :stuck_out_tongue_winking_eye:  
giphy 1

File: app/models/gem_info.rb [hunk 1]
```diff
@@ -44,22 +40,39 @@ def self.versions_after(date)
...
+
+    map_gem_version query
+  end
+
+  def self.map_gem_version(query)
     sanitize_sql = ActiveRecord::Base.send(:sanitize_sql_array, query)
     gems = ActiveRecord::Base.connection.execute(sanitize_sql)
 
     gems.map do |gem|
+      version_checksum = Version._sha256_hex(gem['sha256'])
```
Comment (segiddins) [inline app/models/gem_info.rb]: changing it to hex is unnecessary, the client can handle both hex and base64 https://github.com/bundler/bundler/blob/master/lib/bundler/rubygems_gem_installer.rb#L44

**Comment by segiddins** [comment]
```diff
diff --git a/app/models/gem_info.rb b/app/models/gem_info.rb
index 1fde426..576482f 100644
--- a/app/models/gem_info.rb
+++ b/app/models/gem_info.rb
@@ -41,38 +41,41 @@ class GemInfo
                     v.yanked_at > ?)
               ORDER BY date, number, platform, name", date, date]
 
-    map_gem_version query
+    map_gem_versions execute_raw_sql(query).map {|v| [v['name'], v] }
   end
 
-  def self.compact_index_public_versions(date)
-    query = ["SELECT r.name, v.created_at, v.sha256, v.info_checksum, v.number, v.platform
+  def self.compact_index_public_versions
+    query = ["SELECT r.name, v.indexed, COALESCE(v.yanked_at, v.created_at) as stamp, v.sha256, COALESCE(v.yanked_info_checksum, v.info_checksum) as info_checksum, v.number, v.platform
               FROM rubygems AS r, versions AS v
-              WHERE v.rubygem_id = r.id AND
-                    v.indexed is true AND
-                    v.created_at > ?
-              ORDER BY v.created_at, v.number, v.platform, r.name", date]
+              WHERE v.rubygem_id = r.id
+              ORDER BY r.name, stamp, v.number, v.platform"]
 
-    map_gem_version query
+    map_gem_versions(execute_raw_sql(query).group_by {|v| v['name'] }.each do |_, versions|
+      info_checksum = versions.last['info_checksum']
+      versions.select! {|v| v['indexed'] == 't'.freeze }
+      versions.each {|v| v['info_checksum'] = info_checksum }
+    end)
   end
 
-  def self.map_gem_version(query)
-    sanitize_sql = ActiveRecord::Base.send(:sanitize_sql_array, query)
-    gems = ActiveRecord::Base.connection.execute(sanitize_sql)
+  def self.execute_raw_sql(query)
+    sanitized_sql = ActiveRecord::Base.send(:sanitize_sql_array, query)
+    ActiveRecord::Base.connection.execute(sanitized_sql)
+  end
+  private_class_method :execute_raw_sql
 
-    gems.map do |gem|
-      version_checksum = Version._sha256_hex(gem['sha256'])
-      CompactIndex::Gem.new(gem['name'], [
+  def self.map_gem_versions(versions_by_gem)
+    versions_by_gem.map do |gem_name, versions|
+      CompactIndex::Gem.new(gem_name, Array(versions).map do |version|
                               CompactIndex::GemVersion.new(
-                                gem['number'],
-                                gem['platform'],
-                                version_checksum,
-                                gem['info_checksum']
+                                version['number'],
+                                version['platform'],
+                                version['sha256'],
+                                version['info_checksum']
                               )
-                            ])
+                            end)
     end
   end
-
-  private_class_method :map_gem_version
+  private_class_method :map_gem_versions
 
   private
 
diff --git a/lib/tasks/compact_index.rake b/lib/tasks/compact_index.rake
index 7129596..148383d 100644
--- a/lib/tasks/compact_index.rake
+++ b/lib/tasks/compact_index.rake
@@ -76,20 +76,7 @@ namespace :compact_index do
     file_path = Rails.application.config.rubygems['versions_file_location']
     versions_file = CompactIndex::VersionsFile.new file_path
 
-    gems = GemInfo.compact_index_public_versions(Time.at(0).utc.to_datetime)
-    gems = gems.group_by(&:name)
-    gems = gems.map do |name, compact_index_gems|
-      versions = compact_index_gems.flat_map(&:versions)
-
-      # Ensure we set the info checksums to be that of the last updated version
-      # because the last version could be yanked
-      info_checksum = versions.last.info_checksum
-
-      # Set all versions' info_checksum to work around https://github.com/bundler/compact_index/pull/20
-      versions.each { |v| v.info_checksum = info_checksum }
-
-      CompactIndex::Gem.new(name, versions)
-    end
+    gems = GemInfo.compact_index_public_versions
 
     versions_file.create gems
   end
diff --git a/test/unit/update_versions_file_test.rb b/test/unit/update_versions_file_test.rb
index b1d6e51..264ec8e 100644
--- a/test/unit/update_versions_file_test.rb
+++ b/test/unit/update_versions_file_test.rb
@@ -80,7 +80,8 @@ class UpdateVersionsFileTest < ActiveSupport::TestCase
           created_at:    3.minutes.ago,
           yanked_at:     1.minute.ago,
           number:        "0.0.2",
-          info_checksum: "sd12q")
+          info_checksum: "sd12q",
+          yanked_info_checksum: "qw212r")
         Rake::Task["compact_index:update_versions_file"].invoke
       end
 
@@ -89,6 +90,35 @@ class UpdateVersionsFileTest < ActiveSupport::TestCase
         assert_equal expected_output, @tmp_versions_file.readlines[2]
       end
     end
+
+    context "when the yanked version isn't the latest version" do
+      setup do
+        create(:version,
+          rubygem:       @rubygem,
+          created_at:    5.seconds.ago,
+          number:        "0.1.1",
+          info_checksum: "zqw212r")
+        create(:version,
+          indexed:       false,
+          rubygem:       @rubygem,
+          created_at:    4.seconds.ago,
+          yanked_at:     2.seconds.ago,
+          number:        "0.1.2",
+          info_checksum: "zsd12q",
+          yanked_info_checksum: "zab45d")
+        create(:version,
+          rubygem:       @rubygem,
+          created_at:    3.seconds.ago,
+          number:        "0.1.3",
+          info_checksum: "zrt13y")
+        Rake::Task["compact_index:update_versions_file"].invoke
+      end
+
+      should "not include yanked version" do
+        expected_output = "rubyrubyruby 0.1.1,0.1.3 zab45d\n"
+        assert_equal expected_output, @tmp_versions_file.readlines[2]
+      end
+    end
   end
 
   context "multiple gems" do
```

**Comment by sonalkr132** [comment]
```diff
-  def self.compact_index_public_versions(date)
-    query = ["SELECT r.name, v.created_at, v.sha256, v.info_checksum, v.number, v.platform
+  def self.compact_index_public_versions
+    query = ["SELECT r.name, v.indexed, COALESCE(v.yanked_at, v.created_at) as stamp, v.sha256, COALESCE(v.yanked_info_checksum, v.info_checksum) as info_checksum, v.number, v.platform
               FROM rubygems AS r, versions AS v
-              WHERE v.rubygem_id = r.id AND
-                    v.indexed is true AND
-                    v.created_at > ?
-              ORDER BY v.created_at, v.number, v.platform, r.name", date]
+              WHERE v.rubygem_id = r.id
+              ORDER BY r.name, stamp, v.number, v.platform"]

-    map_gem_version query
+    map_gem_versions(execute_raw_sql(query).group_by {|v| v['name'] }.each do |_, versions|
+      info_checksum = versions.last['info_checksum']
+      versions.select! {|v| v['indexed'] == 't'.freeze }
+      versions.each {|v| v['info_checksum'] = info_checksum }
+    end)
```

Can you please explain why do we need `COALESCE(v.yanked_at, v.created_at) as stamp`, `COALESCE(v.yanked_info_checksum, v.info_checksum) as info_checksum` and `versions.select! {|v| v['indexed'] == 't'.freeze `? We don't have it here or on bundler-api.  Doesn't new version file only need indexed versions?

```diff
+      CompactIndex::Gem.new(gem_name, Array(versions).map do |version|
```
why  `Array(versions)`? It would have following effect in `compact_index_versions`:
```ruby
Array({:a => "a", :b => "b"}) #=> [[:a, "a"], [:b, "b"]]
```

**Comment by segiddins** [comment]
The `coalesce` is to allow us to make sure we have the _last_ info checksum, even if that comes from a yanked version

`Array(versions)` was for the existing method, so it didn't have to pass in an array

**Comment by sonalkr132** [comment]
> The coalesce is to allow us to make sure we have the last info checksum, even if that comes from a yanked version

Gotcha :+1: 

> Array(versions) was for the existing method, so it didn't have to pass in an array

Did you mean `Array.new(versions)`? Please check last example [here](https://ruby-doc.org/core-2.2.3/Array.html#class-Array-label-Creating+Arrays).

**Comment by segiddins** [comment]
No, `Array(versions)` will coalesce to an array if necessary

File: test/unit/update_versions_file_test.rb [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        create(:version,
+          rubygem:       @rubygem,
+          created_at:    3.seconds.ago,
+          number:        "0.1.3",
+          info_checksum: "zrt13y")
+        Rake::Task["compact_index:update_versions_file"].invoke
+      end
+
+      should "not include yanked version" do
+        expected_output = "rubyrubyruby 0.1.1,0.1.3 zrt13y\n"
```
Comment (sonalkr132) [inline test/unit/update_versions_file_test.rb]: getting there.. hold on! I need to understand your diff as well.

**Comment by sonalkr132** [comment]
@segiddins It is :white_check_mark:  now. Thank you for your help. Let me know if it needs anymore change.

**Comment by segiddins** [comment]
Looks good to me!

**Comment by segiddins** [comment]
@indirect r?

**Comment by indirect** [comment]
LGTM.

**Pull request merged**